peps/pep-0742.rst

```diff
@@ -0,0 +1,305 @@
+
+Type narrowing functions must accept at least one positional argument. The type
+narrowing behavior is applied to the first positional argument passed to
+the function. The function may accept additional arguments, but they are
+not affected by type narrowing.
+
+Type narrowing behavior
+-----------------------
+
+To specify the behavior of ``TypeNarrower``, we use the following terminology:
+
+* I = ``TypeNarrower`` input type
+* R = ``TypeNarrower`` return type
+* A = Type of argument passed to type narrowing function (pre-narrowed)
+* NP = Narrowed type (positive; used when ``TypeNarrower`` returned True)
+* NN = Narrowed type (negative; used when ``TypeNarrower`` returned False)
+
+.. code-block:: python
+
+    def narrower(x: I) -> TypeNarrower[R]: ...
+
+    def func1(val: A):
+        if narrower(val):
+            assert_type(val, NP)
+        else:
+            assert_type(val, NN)
+
+The return type ``R`` must be :ref:`consistent with <pep-483-gradual-typing>` ``I``. The type checker should
+emit an error if this condition is not met.
+
+Formally, type *NP* should be narrowed to :math:`A \land R`, the intersection of *A* and *R*, and type *NN* should be narrowed to
+:math:`A \land \neg R`, the intersection of *A* and the complement of *R*.
+In practice, the theoretic types for strict type guards cannot be expressed
+precisely in the Python type system. Type checkers should fall back on
+practical approximations of these types. As a rule of thumb, a type checker
+should use the same type narrowing logic -- and get results that are consistent
+with -- its handling of :py:func:`isinstance`. This guidance allows for changes and
+improvements if the type system is extended in the future.
+
+In particular, if the argument type ``A`` is a union type, ``NP`` may
+be narrowed to those elements of the union that are consistent with ``R``,
+and ``NN`` may be narrowed to those elements of the union that are
+not subtypes of ``R``.
+
+Examples
+--------
+
+Type narrowing is applied in both the positive and negative case::
+
+    from typing import TypeNarrower, assert_type
+
+    def is_str(x: object) -> TypeNarrower[str]:
+        return isinstance(x, str)
+
+    def f(x: str | int) -> None:
+        if is_str(x):
+            assert_type(x, str)
+        else:
+            assert_type(x, int)
+
+Additional type narrowing is applied to the positive case if applicable::
+
+    from collections.abc import Awaitable
+    from typing import Any, TypeNarrower, assert_type
+
+    def isawaitable(x: object) -> TypeNarrower[Awaitable[Any]]:
+        return inspect.isawaitable(x)
+
+    def f(x: Awaitable[int] | int) -> None:
+        if isawaitable(x):
+            assert_type(x, Awaitable[int])
+        else:
+            assert_type(x, int)
+
+It is an error to narrow to a type that is not consistent with the input type::
+
+    from typing import TypeNarrower
+
+    def is_str_list(x: list[object]) -> TypeNarrower[list[str]]:  # Type checker error
```

Comment:
This example has the advantage of being something someone might actually try to write, but the reason for that is that people often don't understand/remember the invariance of mutable container types, and that also makes this example more likely to be confusing. Would it be simpler to not involve variance here and give a more obviously inconsistent example, like `str` vs `int`? Since the main point is that consistency is required.

Reply:
Agreed, will change.